# Direction 1: Creating the "Price_label" and "Busiest_day" columns

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#make path
path = r'C:\Users\praha\OneDrive\Documents\Python\02-2023 Instachart Basket Analysis'

In [3]:
path

'C:\\Users\\praha\\OneDrive\\Documents\\Python\\02-2023 Instachart Basket Analysis'

In [4]:
#import dataframe 'ords_prods_merged'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'orders_products_merged.pkl'))

## Create the 'Price_label' column

In [5]:
#create the 'Price_label' column using loc()
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_label'] = 'High-range product'

In [6]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_label'] = 'Mid-range-product'

In [7]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_label'] = 'Low-range product'

In [8]:
#check the frequency of values
ords_prods_merge['price_label'].value_counts(dropna = False)

Mid-range-product     21860860
Low-range product     10126321
High-range product      417678
Name: price_label, dtype: int64

## Create 'busiest_day' column

In [16]:
#check the frequency of every value in 'orders_day_of_week' column
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [10]:
#create the 'Busiest_day' columns with for-loops
result=[]

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('least busy')
    else:
        result.append('regularly busy')

In [11]:
result

['regularly busy',
 'regularly busy',
 'Busiest day',
 'regularly busy',
 'least busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'least busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'least busy',
 'regularly busy',
 'regularly busy',
 'Busiest day',
 'regularly busy',
 'regularly busy',
 'least busy',
 'regularly busy',
 'Busiest day',
 'regularly busy',
 'Busiest day',
 'least busy',
 'Busiest day',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'regularly busy',
 'Bus

In [12]:
ords_prods_merge['busiest_day'] = result

In [17]:
#check the frequency of values
ords_prods_merge['busiest_day'].value_counts(dropna = False)

regularly busy    22416875
Busiest day        6204182
least busy         3783802
Name: busiest_day, dtype: int64

In [14]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     7.0                 10          0   both   
4                     9.0                 11          1   both   

         price_label     busiest_day  
0  Mid-range-product  regularly busy  
1  Mid-range-product  regularly busy  
2  Mid-range-product     Busiest day  
3  Mid-range-product  regularly busy  
4  Mid-range-product      least busy

# Direction 2: Create the "busiest_days" column

In [18]:
#check the frequency of every value in 'orders_day_of_week' column
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [33]:
#create the 'Busiest_days' columns with for-loops
result=[]

for value in ords_prods_merge['orders_day_of_week']:
    if value == 0:
        result.append('Busiest days')
    elif value == 1:
        result.append('Busiest days')
    elif value == 3:
        result.append('Slowest days')
    elif value == 4:
        result.append('Slowest days')
    else:
        result.append('regularly busy')

In [27]:
ords_prods_merge['busiest_days'] = result

# Direction 3: Check the values and note any observations

In [28]:
#check the frequency of values
ords_prods_merge['busiest_days'].value_counts(dropna = False)

regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

Observations: The "regularly busy" category is still the highest, but the number decreased from 22416875 to 12916111. This happened due to the merging of saturday (0) and sunday (1) into 'busiest days' and the merging of wednesday (3) and thursday (4) into 'slowest days'

# Direction 4: Create the "busiest_period_of_day" column

In [30]:
#check the value of "order_hour_of_day" column
ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

Classification:
Most orders = 10 - 16;
Average orders = 7 - 9 & 17 - 23;
Fewest orders = 0 - 6

In [34]:
ords_prods_merge['order_hour_of_day'].shape

(32404859,)

In [35]:
#create 'busiest_period_of_day' column using loc()
ords_prods_merge.loc[(ords_prods_merge['order_hour_of_day'] <= 16) & (ords_prods_merge['order_hour_of_day'] >= 10), 'busiest_period_of_day'] = 'Most Orders'

In [36]:
ords_prods_merge.loc[(ords_prods_merge['order_hour_of_day'] <= 9) & (ords_prods_merge['order_hour_of_day'] >= 7), 'busiest_period_of_day'] = 'Average Orders'

In [37]:
ords_prods_merge.loc[(ords_prods_merge['order_hour_of_day'] <= 23) & (ords_prods_merge['order_hour_of_day'] >= 17), 'busiest_period_of_day'] = 'Average Orders'

In [38]:
ords_prods_merge.loc[(ords_prods_merge['order_hour_of_day'] <= 6) & (ords_prods_merge['order_hour_of_day'] >= 0), 'busiest_period_of_day'] = 'Fewest Orders'

# Direction 5: Print the Frequency for 'busiest_period_of_day' column

In [39]:
#check the frequency of values
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

Most Orders       18663868
Average Orders    12854170
Fewest Orders       886821
Name: busiest_period_of_day, dtype: int64

In [40]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     7.0                 10          0   both   
4                     9.0                 11          1   both   

         price_label     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range-product  regularly busy  regularly busy           Most Orders  
1  Mid-range-product  regularly busy  regularly busy        Average Orders  
2  Mid-range-product     Busiest day    Busiest days        Average Orders  
3  Mid-range-product  regularly busy    Slowest days           Most Orders  
4  Mid-range-product      least busy    Slowest days        Average Orders

# Direction 7: Export the dataframe

In [41]:
#export the data to "ords_prods_merge.pkl"
ords_prods_merge.to_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'ords_prods_variables_derived.pkl'))